substrate/frame/migrations/src/lib.rs

```diff
@@ -0,0 +1,574 @@
+	/// remaining weight that can be consumed.
+	fn exec_migration(
+		meter: &mut WeightMeter,
+		mut cursor: ActiveCursorOf<T>,
+		is_first: bool,
+	) -> Option<ControlFlow<ActiveCursorOf<T>, ActiveCursorOf<T>>> {
+		let Some(id) = T::Migrations::nth_id(cursor.index) else {
+			Self::deposit_event(Event::UpgradeCompleted);
+			Cursor::<T>::kill();
+			T::OnMigrationUpdate::completed();
+			return None;
+		};
+		let bounded_id: Result<IdentifierOf<T>, _> = id.try_into();
+		let Ok(bounded_id) = bounded_id else {
+			defensive!("The integrity check ensures that all identifiers' MEL bound fits into CursorMaxLen; qed.");
+			Self::upgrade_failed(Some(cursor.index));
+			return None
+		};
+
+		if Historic::<T>::contains_key(&bounded_id) {
+			Self::deposit_event(Event::MigrationSkippedHistoric { index: cursor.index });
+			cursor.advance(System::<T>::block_number());
+			return Some(ControlFlow::Continue(cursor))
+		}
+
+		let max_steps = T::Migrations::nth_max_steps(cursor.index);
+		let next_cursor = T::Migrations::nth_transactional_step(
+			cursor.index,
+			cursor.inner_cursor.clone().map(|c| c.into_inner()),
+			meter,
+		);
+		let Some((max_steps, next_cursor)) = max_steps.zip(next_cursor) else {
+			defensive!("nth_id returned Some(_). Therefore all other nth_* calls must return Some(_) as well; qed");
+			Self::upgrade_failed(Some(cursor.index));
+			return None
+		};
+
+		let blocks = System::<T>::block_number().saturating_sub(cursor.started_at);
+		match next_cursor {
+			Ok(Some(next_cursor)) => {
+				let Ok(bound_next_cursor) = next_cursor.try_into() else {
+					defensive!("The integrity check ensures that all cursors' MEL bound fits into CursorMaxLen; qed");
+					Self::upgrade_failed(Some(cursor.index));
+					return None
+				};
+
+				Self::deposit_event(Event::MigrationAdvanced { index: cursor.index, blocks });
+				cursor.inner_cursor = Some(bound_next_cursor);
+
+				// We only progress one step per block.
+				if max_steps.map_or(false, |max| blocks > max.into()) {
+					Self::deposit_event(Event::MigrationFailed { index: cursor.index, blocks });
+					Self::upgrade_failed(Some(cursor.index));
+					None
+				} else {
+					// A migration cannot progress more than one step per block, we therefore break.
+					Some(ControlFlow::Break(cursor))
+				}
+			},
+			Ok(None) => {
+				Self::deposit_event(Event::MigrationCompleted { index: cursor.index, blocks });
+				Historic::<T>::insert(&bounded_id, ());
+				cursor.advance(System::<T>::block_number());
+				return Some(ControlFlow::Continue(cursor))
+			},
+			Err(SteppedMigrationError::InsufficientWeight { required }) => {
+				if is_first || required.any_gt(meter.limit()) {
+					Self::deposit_event(Event::MigrationFailed { index: cursor.index, blocks });
+					Self::upgrade_failed(Some(cursor.index));
+					None
+				} else {
+					// Hope that it gets better in the next block.
+					Some(ControlFlow::Break(cursor))
```

Comment:
This seems a bit weird that `InsufficientWeight` could be emitted when `required < meter.limit()`? Does that indicate a buggy migration?

Reply:
There is the case that another migration already finished in the current block and ate up ie. 50% of the block.  
Then the second migration realizes that it does not have enough weight and errors, since it needs 75% (as example).